container/credentials.bzl

```diff
@@ -0,0 +1,118 @@
+load("//container:providers.bzl", "RegistryCredentialInfo")
+
+def write_docker_config_json_action(ctx, registry, credential, config_json = None):
+    """Create an action that writes a docker config file.
+
+    Args:
+      ctx: the rule ctx object
+      registry: the name of the registry to which the credential should be
+      associated
+      credential: a RegistryCredentialInfo struct
+      config_json: path to the file where the config.json should be written.
+      Defaults to %{credential.name}/.docker/config.json.
+
+    Return: The config_json file 
+    """
+    if credential.type != "basic":
+        fail("Unsupported docker credential type: %s" % credential.type)
+    if not config_json:
+        config_json = ctx.actions.declare_file("%s/.docker/config.json" % credential.name)
+    # NOTE: ctx.actions.write does not work here as we need the
+    # use_default_shell_env feature of the run_shell action.
+    ctx.actions.run_shell(
```

Comment:
you should use a template file + expand_template to avoid using shell to create a file (which would only work on environments that have bash)  https://docs.bazel.build/versions/master/skylark/lib/actions.html#expand_template

Reply:
Yes that was my initial thought.  However, in order to use `--action_env` you MUST use a `run` or `run_shell` action, which has the `use_default_shell_env` feature.

Reply:
I really dislike having this code here. We're in process of fixing any instances of usage of tools that will block these rules from working with remote execution or with windows. The way to avoid this hard restriction (--action_env you MUST use a run or run_shell), I think we could use ctx.var (https://docs.bazel.build/versions/master/skylark/lib/ctx.html#var). The main problem with that is that we would not get the env var to be part of the action key. In any case, it should at least be possible to use templates to generate the file with the contents you are generating here using format, and then use run_shell to execute the file produced by the template function.
However, I think we need to hash out the issue about how to do this for container_pull and that will likely require change how were doing this with --action_env. I don't want there to be two different paths to achieve the same thing for container_pull/push.